icu4c/source/common/rbbi_cache.cpp

```diff
@@ -341,41 +341,82 @@ UBool RuleBasedBreakIterator::BreakCache::populateNear(int32_t position, UErrorC
     }
     U_ASSERT(position < fBoundaries[fStartBufIdx] || position > fBoundaries[fEndBufIdx]);
 
-    // Find a boundary somewhere in the vicinity of the requested position.
-    // Depending on the safe rules and the text data, it could be either before, at, or after
-    // the requested position.
-
+    // Add boundaries to the cache near the specified position.
+    // The given position need not be a boundary itself.
+    // The input position must be within the range of the text, and
+    // on a code point boundary.
+    // If the requested position is a break boundary, leave the iteration
+    // position on it.
+    // If the requested position is not a boundary, leave the iteration
+    // position on the preceding boundary and include both the
+    // preceding and following boundaries in the cache.
+    // Additional boundaries, either preceding or following, may be added
+    // to the cache as a side effect.
 
     // If the requested position is not near already cached positions, clear the existing cache,
     // find a near-by boundary and begin new cache contents there.
 
-    if ((position < fBoundaries[fStartBufIdx] - 15) || position > (fBoundaries[fEndBufIdx] + 15)) {
-        int32_t aBoundary = 0;
-        int32_t ruleStatusIndex = 0;
-        if (position > 20) {
-            int32_t backupPos = fBI->handleSafePrevious(position);
-
-            if (backupPos > 0) {
-                // Advance to the boundary following the backup position.
-                // There is a complication: the safe reverse rules identify pairs of code points
-                // that are safe. If advancing from the safe point moves forwards by less than
-                // two code points, we need to advance one more time to ensure that the boundary
-                // is good, including a correct rules status value.
-                //
-                fBI->fPosition = backupPos;
-                aBoundary = fBI->handleNext();
-                if (aBoundary <= backupPos + 4) {
-                    // +4 is a quick test for possibly having advanced only one codepoint.
-                    // Four being the length of the longest potential code point, a supplementary in UTF-8
-                    utext_setNativeIndex(&fBI->fText, aBoundary);
-                    if (backupPos == utext_getPreviousNativeIndex(&fBI->fText)) {
-                        // The initial handleNext() only advanced by a single code point. Go again.
-                        aBoundary = fBI->handleNext();   // Safe rules identify safe pairs.
-                    }
+    int32_t aBoundary = -1;
+    int32_t ruleStatusIndex = 0;
+    bool retainCache = false;
+    if ((position > fBoundaries[fStartBufIdx] - 15) && position < (fBoundaries[fEndBufIdx] + 15)) {
+        // Requested position is near the existing cache. Retain it.
+        retainCache = true;
+    } else if (position <= 20) {
```

Comment:
What is this magic number 20 here? Could we add a constant for that and add a note about how we pick that number?

Reply:
Sounds reasonable -- I am ok with keeping magic numbers, together with a comment about them.

Reply:
so... 
1. how do you come out with the guess of each.
2. Is there a requirement to keep some of these guessed magic number in sync? so if I later change one 15 to 18, I have to change another 15 to 18 together or otherwise it will cause bug?  for example, there aer two "15" in line 362, if we later change one of them but not both of them to 18 , would that cause bug? How about the 15 in line 375?  I cannot tell the "relationship" between these "15"s . Are they all independent or are some of them need to be the same value even currently they are all 15? any constrain of these guess? could we later guess 14 or 16 and won't cause problem?

Reply:
These numbers all represent a concept of "close". Is a requested position close to existing boundary cache contents, or close to a known boundary, like the start of text. If close, the code will prepend or append to existing cache contents to pick up the needed content. If not close, existing cache contents are dumped and a new cache is started around the needed position. The cache can only hold a contiguous set of boundary positions. It can't handle boundaries from discontinuous regions of the input text

There is no required relationship between any of "close" numbers. Any of them can be adjusted independently of the others. Any values will work - the break iterator will still produce correct results. What will change is the performance.

Choosing optimum "close" numbers will depend on the scenario being measured. And probably the break iterator type.  It's going to be tricky to do. For now, I think we just want to verify that the various tests and code paths do what the comments say, retaining the cache when its content is "close" and discarding it when not. You don't even need to worry about off-by-one errors.

I added the new random access test specifically to tickle all of the code paths through this function. And stepped through a bunch of them in the debugger to verify that the code was doing what I expected. There's no other obvious easy way to test it; the break iterator still produces correct results if the cache management is broken, just with worse performance.

Reply:
>There is no required relationship between any of "close" numbers. Any of them can be adjusted independently of the others.
ok. My concern is such "no required connection" is not clear to the reader of the code 10 years later. How could we land the code in a way that clearly indicating one 15 not need to be the same value as another 15 in the code. Should we add 
// The following value do NOT need to be the same. There is no required relationship between any of 
// these radom guess. They could be changed independently even if they currently share the same value.
constexpr int32_t kRandomGuess1  = 15;
constexpr int32_t kRandomGuess2 = 15;
constexpr int32_t kRandomGuess3  = 15;
constexpr int32_t kRandomGuess4 = 15;
constexpr int32_t kRandomGuess5 = 4;
constexpr int32_t kRandomGuess6 = 20;

and use them in different spots?
(or maybe come out same better name for these random guess.)